Improve testing coverage

bratislavSyrmia opened the pull request:
Create tests for:

  1) -o option
  2) -l option
  3) -h option
  4) -s option
  5) -a option
  6) presence of debug info in object file generated by -s option
  7) `SIGQUIT`

Basically this resolves issue: https://github.com/ROCm/rocr_debug_agent/issues/15

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-p -o output_log.txt"
+    # Start process without capturing output
+    p = Popen(['./rocm-debug-agent-test', '1'])
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-p"
```
probably use the pattern existing before to set/restore the env variable

```python
        with unittest.mock.patch.dict(os.environ, {"ROCM_DEBUG_AGENT_OPTIONS":  
                                                   f"-o out_log.txt"}):
```

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-p -o output_log.txt"
+    # Start process without capturing output
+    p = Popen(['./rocm-debug-agent-test', '1'])
```
the idea seems to re-run test 1, but with this new argument.  Can't we refactor a bit so the actual logic to check the desired output is shared?

bratislavSyrmia replied:
I can do it if you want me to, but I would prefer to keep it separated because:
1) Output is not 100% same, there are some things that arent written to file, for example
`HSA_STATUS_ERROR_EXCEPTION: An HSAIL operation resulted in a hardware exception`
2) This keeps tests for each option separate, so for example if the trap test works fine, but there is en error when writing to file, it would be easier to spot where the problem is

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+
+    all_output_string_found = True
+    for check_str in check_list:
+        pattern = re.compile(check_str)
```
(this comment applies to already existing code)

There is a "TODO: use regular expressions instead of strings" comment above, but the elements of `check_list` are treated as regex to compile and match against the output.

The comment is probably not accurate.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    # check output string
+    all_output_string_found = True
+    for check_str in check_list:
+        pattern = re.compile(check_str)
```
same remark regarding the TODO above.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    err_str = err.decode('utf-8')
+
+    print(out_str)
+    print(err_str)
```
Was this for debug?  L280-284 do dump the same information on failure.

As the rest of the file dumps of failure, we probably want to keep the same pattern.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    check_list = ['ROCdebug-agent usage',
+                  ]
+
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-h"
```
Same remark, probably prefer unittest.mock.patch.dict to handle the update/restore this value.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+def check_test_7():
+    print("Starting rocm-debug-agent test 7")
+
+    #TODO: use regular expressions instead of strings
```
Same remark regarding the TODO.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    check_list = ['rocm-dbgapi',
+                  ]
+
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-l info"
```
Same remark regarding unittest.mock.patch.dict.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    err_str = err.decode('utf-8')
+
+    print(out_str)
+    print(err_str)
```
Same remark regarding the duplicate prints.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    print("Starting rocm-debug-agent test 7")
+
+    #TODO: use regular expressions instead of strings
+    check_list = ['rocm-dbgapi',
```
probably `'^rocm-dbgapi: ` to try to be a tiny bit more specific.

Most of the rest of the output would depend on dbgapi's internal implementation, so could be subject to evolve.  Using the "trace" log-level would trace function call to stable ABI we can rely on.  That would require a small change in the debug agent to expose this log level option:

```diff
diff --git a/src/debug_agent.cpp b/src/debug_agent.cpp
index ea7756c..3bdbc1e 100644
--- a/src/debug_agent.cpp
+++ b/src/debug_agent.cpp
@@ -1519,6 +1519,8 @@ OnLoad (void *table, uint64_t runtime_version, uint64_t failed_tool_count,
             set_log_level (log_level_t::warning);
           else if (argument == "error")
             set_log_level (log_level_t::error);
+          else if (argument == "trace")
+            set_log_level (log_level_t::trace);
           else
             print_usage ();
           break;
diff --git a/src/logging.cpp b/src/logging.cpp
index 2111475..e26dc91 100644
--- a/src/logging.cpp
+++ b/src/logging.cpp
@@ -90,6 +90,9 @@ set_log_level (log_level_t level)
     case log_level_t::warning:
       amd_dbgapi_set_log_level (AMD_DBGAPI_LOG_LEVEL_WARNING);
       break;
+    case log_level_t::trace:
+      amd_dbgapi_set_log_level (AMD_DBGAPI_LOG_LEVEL_TRACE);
+      break;
     case log_level_t::error:
       amd_dbgapi_set_log_level (AMD_DBGAPI_LOG_LEVEL_FATAL_ERROR);
       break;
diff --git a/src/logging.h b/src/logging.h
index 1bd3d1c..fb986a6 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -46,8 +46,10 @@ enum class log_level_t
   warning = 2,
   /* Print error, warning, and info messages.  */
   info = 3,
+  /* Print errors, warning, trace and info messages.  */
+  trace = 4,
   /* Print error, warning, info, and verbose  messages.  */
-  verbose = 4
+  verbose = 5
 };
 
 extern log_level_t log_level;
```

then we have a trace of calls made by the debug-agent itself, so should be pretty stable:

```
rocm-dbgapi: amd_dbgapi_initialize (0x714138b456c0) {
rocm-dbgapi:    library info: file_name="/opt/rocm/lib/librocm-dbgapi.so.0", build_info=0.77.2-rocm-dkms-no-npi-hipclang-15809
rocm-dbgapi: } = STATUS_SUCCESS
rocm-dbgapi: amd_dbgapi_process_attach (client_process_id=0x7140375f7434, process_id=0x7140375f7478) {
[…]
```

This could be overkill.

bratislavSyrmia replied:
IMPO, this makes sense, but I would recommend adding this as a separate PR. Please let me know what you thinik

lancesix replied:
I am happy with that to be addressed in a separate PR.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-h"
+    p = Popen(['./rocm-debug-agent-test', '1'], stdout=PIPE, stderr=PIPE)
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-p"
```
The `-p` as a default value has been removed in

    commit bac576a0df95487f1f1b5cf161a4cefa8fbb060a
    Date:   Tue Apr 1 10:35:28 2025 +0100

        Prevent VGPR_DEALLOC in est/vector_add_memory_fault.cpp

Using the unittest.mock approach would avoid incorrect restoration of the previous value.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    print("Starting rocm-debug-agent test 8")
+
+    # Set up environment to save code objects in the current directory
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "-p -s ./"
```
See previous remarks.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    for f in glob("*file____*"):
+        os.remove(f)
+    for f in glob("*memory__*"):
+        os.remove(f)
```
test_4 (which arguably should have an actual name) uses a temporary directory to avoid using the base build dir.  Did you go this way of purpose?

If the intent is to keep file on error for manual check, one could do

```python
with tempfile.TemporaryDirectory(suffix=..., delete=False) as d:
    # work with d
    if all_ok:
        d.cleanup()
        return True
    return False
```

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+        objdump = Popen(["objdump", "-g", f], stdout=PIPE, stderr=PIPE)
+        out, _ = objdump.communicate()
+        if b".debug_info" in out:
+            return True
```
I am not sure this tests much of the debug agent itself.

Having debug information or not is a property of the application under test, not of the agent itself.  Running locally, I see that most copied code objects debug information, but not all

```
$ for i in *__tmp* *_memory_* ;  do echo $i; readelf -S $i|grep debug_info; done
1_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_32768_size_34216
  [12] .debug_info       PROGBITS         0000000000000000  000012dc
2_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_73728_size_30688
  [12] .debug_info       PROGBITS         0000000000000000  00000a84
3_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_110592_size_34184
  [13] .debug_info       PROGBITS         0000000000000000  0000108c
4_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_151552_size_34856
  [12] .debug_info       PROGBITS         0000000000000000  000014be
5_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_192512_size_33304
  [13] .debug_info       PROGBITS         0000000000000000  00000f0d
6_memory___756753_offset_0x33ae01e0_size_40048
```

One could check that the code object do match some that where from the input code object to ensure they are valid.

```
$ roc-obj rocm-debug-agent-test
$ sha256sum *_file_* rocm-debug-agent-test\:*|sort
0dedf819d76a4b226af1bed4871dea7758ffc77d420ee0a20d4bc1886fedfabd  1_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_32768_size_34216
0dedf819d76a4b226af1bed4871dea7758ffc77d420ee0a20d4bc1886fedfabd  rocm-debug-agent-test:1.hipv4-amdgcn-amd-amdhsa--gfx1031
68aa7392b3ebe2435e0ce3cdeaeb9f420152d9c98b60f90543df487467331793  2_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_73728_size_30688
68aa7392b3ebe2435e0ce3cdeaeb9f420152d9c98b60f90543df487467331793  rocm-debug-agent-test:2.hipv4-amdgcn-amd-amdhsa--gfx1031
8d6e94756492d2fc2f2e3de2ad68c15c695a8d67647dad5c5419594ba4340ff6  5_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_192512_size_33304
8d6e94756492d2fc2f2e3de2ad68c15c695a8d67647dad5c5419594ba4340ff6  rocm-debug-agent-test:5.hipv4-amdgcn-amd-amdhsa--gfx1031
ca8960d276f2a582c08632790a38b4e3d5e4c81fcfcc8410fceb7a326a24e1e2  3_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_110592_size_34184
ca8960d276f2a582c08632790a38b4e3d5e4c81fcfcc8410fceb7a326a24e1e2  rocm-debug-agent-test:3.hipv4-amdgcn-amd-amdhsa--gfx1031
dfbfea176f448970653a331a03cf971eda146698748388dbcc5d4c8f6bd66a16  4_file____tmp_dbgagent_test_rocm-debug-agent-test_offset_151552_size_34856
dfbfea176f448970653a331a03cf971eda146698748388dbcc5d4c8f6bd66a16  rocm-debug-agent-test:4.hipv4-amdgcn-amd-amdhsa--gfx1031
```

bratislavSyrmia replied:
Should I remove this test then, since it doesn't test anything relevant to us, and test the presence of debug info in the way that I suggested in the comment bellow (editing test 1 or creating one similar to it that also checks for additional debug info)?

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+def check_test_9():
+    print("Starting rocm-debug-agent test 9")
+
+    #TODO: use regular expressions instead of strings
```
Same remark regarding the TODO / regex.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+
+    #TODO: use regular expressions instead of strings
+    check_list = ['wave_1',
+    'wave_2',
```
Indentation seems odd.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    'wave_8',
+                  ]
+
+    os.environ["ROCM_DEBUG_AGENT_OPTIONS"] = "--all"
```
See previous remark.

lancesix commented on test/run-test.py:
```diff
@@ -210,6 +213,231 @@ def check_test_4():
+    err_str = err.decode('utf-8')
+
+    print(out_str)
+    print(err_str)
```
See previous remarks.

lancesix commented on test/sigquit.cpp:
```diff
@@ -0,0 +1,23 @@
+{
+  hipError_t err;
+  sigquit_kern<<<1, 1>>> ();
+  err = hipDeviceSynchronize ();
```
When running outside of the testsuite, this would hang forever.

Should there be a use of  `alarm(2)` to be sure the process will terminate after 30sec or so?

lancesix commented on test/sigquit.cpp:
```diff
@@ -0,0 +1,23 @@
+
+#include <cstdlib>
+#include <string.h>
+#include <string>
```
I don't think any of the cstdlib, string.h or string is needed for this particular test.

lancesix commented on test/print_all_waves.cpp:
```diff
@@ -0,0 +1,26 @@
+#pragma clang optimize off
```
I don't expect this is necessary.

bratislavSyrmia replied:
I think it is. I've compiled this file as a standalone with hipcc, when I pass it the -o0 option it prints all waves, but when I compiled it without that option it only prints the waves that caused the crash. I tried adding some dummy code to avoid optimizing out the other waves, for example each thread writes its global id into some volatile variable, but no success.

lancesix commented on test/print_all_waves.cpp:
```diff
@@ -0,0 +1,26 @@
+
+#include <cstdlib>
+#include <string.h>
+#include <string>
```
None of the above 3 includes seem used.

lancesix commented on test/debug_agent_test.cpp:
```diff
@@ -202,3 +212,48 @@ RunSaveCodeObjectTest ()
+  int deviceCount;
+  hipError_t err = hipGetDeviceCount (&deviceCount);
+  TEST_ASSERT (err == hipSuccess, "hipGetDeviceCount");
+
```
This looks like an extra empty line.

lancesix commented on test/debug_agent_test.cpp:
```diff
@@ -202,3 +212,48 @@ RunSaveCodeObjectTest ()
+  hipError_t err = hipGetDeviceCount (&deviceCount);
+  TEST_ASSERT (err == hipSuccess, "hipGetDeviceCount");
+
+
```
same, probably one empty line is enough.

lancesix reviewed (commented):
Hi,

Thanks for adding this.
I have added  comments, many are the same thing repeated at different places.

lancesix commented:
oh, I think I just got why you added the test regarding the presence of debug info.  I think this comes from this item in #15

```
* Presence or not of debug information
```

This was about testing the different outputs the debug agent can produce when an exception occurs from a code for which we have debug info v.s. from a code for which we do not have debug info.  I.e. check that when running the application build with `-g`, we have more info.

i.e. using the testcase 1, with debug info we have:

```
$ HSA_TOOLS_LIB=/opt/rocm/lib/librocm-debug-agent.so.2 ./rocm-debug-agent-test 1
[…]
/.../test/vector_add_assert_trap.cpp:
55        c[gid] = a[gid] + b[gid] + (lds_check[0] >> 32);
    ...
    0x799a43c19acc <+204>:    v_add3_u32 v2, v3, v2, v4
    0x799a43c19ad4 <+212>:    global_store_dword v[0:1], v2, off

52        if (gid == 0)
    0x799a43c19adc <+220>:    s_andn2_saveexec_b32 s0, s0
    0x799a43c19ae0 <+224>:    s_cbranch_execz 65503  # 0x799a43c19a60 <vector_add_assert_trap(int*, int*, int*)+96>

53          __builtin_trap ();
    0x799a43c19ae4 <+228>:    s_trap 2
 => 0x799a43c19ae8 <+232>:    s_endpgm

```

but building the same thing without debug info would give

```
$ HSA_TOOLS_LIB=/opt/rocm/lib/librocm-debug-agent.so.2 ./rocm-debug-agent-test 1
[…]
Disassembly for function vector_add_assert_trap(int*, int*, int*):
    code object: file:///tmp/dbgagent/test/rocm-debug-agent-test#offset=53248&size=4736
    loaded at: [0x7d2fd5818000-0x7d2fd581bc31]
 => 0x7d2fd5819ae8 <+232>:    s_endpgm
    0x7d2fd5819aec <+236>:    s_code_end
    0x7d2fd5819af0 <+240>:    s_code_end
    0x7d2fd5819af4 <+244>:    s_code_end
    0x7d2fd5819af8 <+248>:    s_code_end
    0x7d2fd5819afc <+252>:    s_code_end
    ...

```

bratislavSyrmia commented:
> oh, I think I just got why you added the test regarding the presence of debug info. I think this comes from this item in #15
> 
> ```
> * Presence or not of debug information
> ```
> 
> This was about testing the different outputs the debug agent can produce when an exception occurs from a code for which we have debug info v.s. from a code for which we do not have debug info. I.e. check that when running the application build with `-g`, we have more info.
> 
> i.e. using the testcase 1, with debug info we have:
> 
> ```
> $ HSA_TOOLS_LIB=/opt/rocm/lib/librocm-debug-agent.so.2 ./rocm-debug-agent-test 1
> […]
> /.../test/vector_add_assert_trap.cpp:
> 55        c[gid] = a[gid] + b[gid] + (lds_check[0] >> 32);
>     ...
>     0x799a43c19acc <+204>:    v_add3_u32 v2, v3, v2, v4
>     0x799a43c19ad4 <+212>:    global_store_dword v[0:1], v2, off
> 
> 52        if (gid == 0)
>     0x799a43c19adc <+220>:    s_andn2_saveexec_b32 s0, s0
>     0x799a43c19ae0 <+224>:    s_cbranch_execz 65503  # 0x799a43c19a60 <vector_add_assert_trap(int*, int*, int*)+96>
> 
> 53          __builtin_trap ();
>     0x799a43c19ae4 <+228>:    s_trap 2
>  => 0x799a43c19ae8 <+232>:    s_endpgm
> ```
> 
> but building the same thing without debug info would give
> 
> ```
> $ HSA_TOOLS_LIB=/opt/rocm/lib/librocm-debug-agent.so.2 ./rocm-debug-agent-test 1
> […]
> Disassembly for function vector_add_assert_trap(int*, int*, int*):
>     code object: file:///tmp/dbgagent/test/rocm-debug-agent-test#offset=53248&size=4736
>     loaded at: [0x7d2fd5818000-0x7d2fd581bc31]
>  => 0x7d2fd5819ae8 <+232>:    s_endpgm
>     0x7d2fd5819aec <+236>:    s_code_end
>     0x7d2fd5819af0 <+240>:    s_code_end
>     0x7d2fd5819af4 <+244>:    s_code_end
>     0x7d2fd5819af8 <+248>:    s_code_end
>     0x7d2fd5819afc <+252>:    s_code_end
>     ...
> ```

Oh I see. Could I just modify test 1 to check for some additional lines then, for example 
"s_andn2_saveexec_b32 s0, s0"
"s_cbranch_execz 65503"
or make some additional test that calls ./rocm-debug-agent-test 1 and checks for those lines?

lancesix commented:
Hi @bratislavSyrmia , sorry for the delay.

In our development model, we follow debugger (GDB)'s practices of working on a patch series before applying them on top of amd-staging.  That is: reviews are to be integrated on the original patches, not as new commits stacked on top of the PR.  Can I let you adjust this PR to integrate your change in the original patches you have?

Thanks.

bratislavSyrmia committed:
test: Add test for -o option

bratislavSyrmia committed:
test: Add test for -h option

bratislavSyrmia committed:
test: Add test for -l option

bratislavSyrmia committed:
test: Add test for checking presence of debug info

bratislavSyrmia committed:
test: Add test for -a option

bratislavSyrmia committed:
test: Add test for SIGQUIT

bratislavSyrmia committed:
test: Add test for debug flags

Add test to check for difference between files compiled with and
without debug flags.

bratislavSyrmia committed:
test: Minor changes to test infrastructure

Removing unnecessary comments from tests.

bratislavSyrmia commented:
@lancesix Thanks a lot for your comments, the changes are squashed into the original commits.

ammallya commented:
Failed to migrate to ROCm/rocm-systems because of merge conflicts please reopen this PR on ROCm/rocm-systems
